### DISCUSSION TITLE
Нотификации

### DISCUSSION THREAD

**Discussion by picolino** [habralab/garnet-team#149] [Идеи]
### Описание
В рамках продукта предусмотрена отправка уведомлений пользователям о следующих событиях:
- Пользователю приходит приглашение о вступлении в команду
- Пользователю приходит оповещение о принятии\отказе о заявке на вступление в команду
- Пользователю-владельцу команды приходит заявка на вступление в команду
- Пользователю-владельцу команды приходит оповещение о принятии\отказе приглашения пользователя о вступлении в команду
- Пользователю-владельцу проекта приходит заявка на вступление в проект от владельца команды
- Пользователю-владельцу команды приходит оповещение о принятии\отказе заявки на вступление в проект.

### Техническая реализация
- Делаем отдельный микросервис нотификаций
- На стороне микросервиса должно присутствовать событие (Event) `SendNotificationCommandMessage` и присутствовать обработчик этого события.
- В отдельную базу данных микросервиса складируем отправляемые уведомления.
- На стороне микросервиса реализуем GraphQL эндпоинты для чтения нотификаций по пользователю и функционал "прочтения" нотификаций
- Текст нотификаций определяется ОТПРАВЛЯЮЩИМ СОБЫТИЕ микросервисом. То есть поля `SendNotificationCommandMessage` должны быть максимально контексто-независимыми. Например, содержать только `Title`, `UserId` и `Body`.

**Comment by gfg7** [comment]
Мне кажется, было бы прикольнее, если в проекте будет разный флоу для переадресации в фронте по клику на уведомление, а не просто статическое уведомление

Например:
+ владельцу команды пришло уведомление, что пользователь отправил заявку на вступление - владелец кликает на уведомление и переходит в карточку команды с открытым списком заявок
+ пришло уведомление, что заявку на вступление в команду одобрили/отклонили - пользователь может кликнуть на уведомление и перейти на карточку команды

Вопрос, как это реализовать на стороне бэка
Добавить в SendNotificationCommandMessage опциональное id сущности
А как фронт определит, что за сущность, куда идти? Задавать тип уведомления, чтобы определить флоу - так мы на себя перетягиваем одеяло, мне кажется

**Reply by picolino** [reply]
Можно добавить ещё тип уведомления (по аналогии с кодом ошибки).

На основании типа уведомления и айдишника сущности фронт будет определять куда перенаправить пользователя

Кажется этой связки будет достаточно. Если нет, то на крайняк можно будет сделать Generic и в зависимости от типа уведомления хранить в нем разные данные, полностью описывающие контекст уведомления.

**Comment by Demorey** [comment]
Предлагаю сделать делать архив отклоненных приглашений/заявок, уведомления не должны храниться с целью поиска сущности, но пользователь сможет посмотреть кто ему отказал или кому отказал он

**Reply by TorinAsakura** [reply]
Может идти от простого? Зачем какие-то архивы или уведомления? Действия с карточкой? Добавь прямо в карточку.

![Снимок экрана 2023-10-31 в 00 58 56](https://github.com/habralab/garnet-team/assets/3940885/dbd2ccc9-262b-480c-b5ad-7d08c86a49d7)

**Reply by Demorey** [reply]
> Может идти от простого? Зачем какие-то архивы или уведомления? Действия с карточкой? Добавь прямо в карточку.
> 
> ![Снимок экрана 2023-10-31 в 00 58 56](https://github.com/habralab/garnet-team/assets/3940885/dbd2ccc9-262b-480c-b5ad-7d08c86a49d7)
> 

Мне кажется что заявки это все таки что то приватное, в ленте пользователя можно как раз отобразить его успехи (подтвержденные заявки/приглашения), а вот неудачи в виде отказов думаю отображать мало кто захотел бы

**Reply by TorinAsakura** [reply]
а чего в заявках приватного если ты потом всё равно появишься в команде в том или ином виде, а команда появится у тебя под профайлом?

**Reply by Demorey** [reply]
Ну я потому и предлагаю не отображать в ленте отказы (т.к. это негатив + их потенциально может быть очень много, в ленте это будет как спам), а просто архивировать их

**Reply by TorinAsakura** [reply]
так я вообще не про то, я про то, что уведомления, которые вы хотели рассылать в письмах - достаточно отображать в ленте, а что отображать это уже вопрос деталей, а я больше про концепт
